### PR TITLE
Implement `fpm publish`

### PR DESCRIPTION
**Pull request by minhqdao**
Implement api for publishing a package to the registry.

- [x] Add new `fpm_publish_settings` and `fpm_cmd_publish` module.
- [x] Add procedure to find system tmp folder.
- [x] Add procedure to execute command line and read output.
- [x] Optionally show request with `--show-form-data` instead of publishing.
- [x] Optionally show package version with `--show-version` instead of publishing.
- [x] Parse license from the manifest.
- [x] Parse token from command.
- [x] Validate data.
- [x] Reject packages containing `git` dependencies.
- [x] Pack and archive using `git archive`.
- [x] Determine available archive methods with `git archive -l` and use available method.
- [x] Send request with form data using `curl`.
- [x] Add tests.

### How to test

1) Run `fpm publish --show-package-version` to print the package version to the console.

2) Run `fpm publish --show-form-data` to print the form data to the console that would otherwise be sent through `fpm publish --token <token>`. If you include the token (`fpm publish --token <token> --show-form-data`), the token will be included in the form data.

3) Get a (test) token from the registry website and run `fpm publish --token <token>` to publish a package to the (test) registry.

### CODE REVIEW

**Review by perazz — APPROVED** [review]
Thanks @minhqdao, looks good to me.

**Review by arteevraina — APPROVED** [review]
@minhqdao Looks good to me.

**Review by henilp105 — APPROVED** [review]
Thanks @minhqdao , Looks good to me.

File: src/fpm_filesystem.F90 [hunk 1]
```diff
@@ -1020,4 +1021,65 @@ subroutine get_home(home, error)
         end if
     end subroutine get_home
 
+    !> Execute command line and return output as a string.
```
Comment (minhqdao) [inline src/fpm_filesystem.F90:1024]: I think this should be built into fortran just as `execute_command_line`. I was searching whether we already have a routine implemented in `fpm` but couldn't find any. Have I missed sth? @awvwgk @everythingfunctional @milancurcic

File: src/fpm_filesystem.F90 [hunk 2]
```diff
@@ -1020,4 +1021,65 @@ subroutine get_home(home, error)
...
+          tmp_dir = tmp_dir//'fpm'; return
+        end if
+
+        tmp_dir = get_env('TEMP', '')
+        if (tmp_dir /= '') then
+          tmp_dir = tmp_dir//'fpm'; return
+        end if
+
+        call fatal_error(error, "Couldn't determine system temporary directory.")
+    end
```
Comment (minhqdao) [inline src/fpm_filesystem.F90:1083]: Same for this actually. @awvwgk @everythingfunctional @milancurcic

**Review by awvwgk — APPROVED** [review]
Looks good to me as well

File: src/fpm_filesystem.F90 (hunk 1 above)
Comment (everythingfunctional) [reply src/fpm_filesystem.F90:1024]: I haven't found any better method of capturing external command output in Fortran. I could potentially mention it to the standards committee.

File: src/fpm_filesystem.F90 (hunk 2 above)
Comment (everythingfunctional) [reply src/fpm_filesystem.F90:1083]: This one is built in.

```Fortran
error stop "Couldn't ..."
```

File: src/fpm_filesystem.F90 (hunk 1 above)
Comment (minhqdao) [reply src/fpm_filesystem.F90:1024]: I think it would be great if we had that.

**Pull request merged**